documentation/trouble_low_catalog_duplication.markdown

```diff
@@ -1,172 +0,0 @@
----
-title: "PuppetDB 3.1 » Troubleshooting » Low Catalog Duplication"
-layout: default
-canonical: "/puppetdb/latest/trouble_low_catalog_duplication"
----
-
-[maintaining_tuning]: ./maintain_and_tune.html
-[configure_vardir]: ./configure.html#vardir
-[global_config]: ./configure.html#global-settings
-[configure_vardir]: ./configure.html#catalog-hash-conflict-debugging
-[configure_catalog_debugging]: ./configure.html#catalog-hash-conflict-debugging
-
-What is Catalog Duplication?
------
-
-Every puppet agent run on a host results in a new catalog sent from
-the puppet master to PuppetDB. Although catalogs can be very large,
-typically there are 1000 or more resources and their relationships in
-each of the catalogs. Storing each of these resources and
-relationships is a fairly I/O-intensive operation on the database used
-by PuppetDB.
-
-Puppet agent runs happen frequently; typically every 60
-minutes. Most of these runs will not result in a catalog with
-different resources or different relationships. Typically, only a
-manually initiated change to Puppet content, such as modification of a
-manifest or a Puppet extension, will update a node's catalog.
-
-PuppetDB stores catalogs with this use case in mind. PuppetDB will SHA1 hash
-the contents of a catalog and will compare that hash with the hash of
-the previously received catalog for that host. Only when these hashes
-don't match will PuppetDB incur the extra cost/overhead of storing the
-new catalog (with all of its resources and edges).
-
-The _catalog duplication rate_ is a measure of how often PuppetDB is able
-to _not_ store a whole catalog. It can be found on the PuppetDB dashboard as a
-percentage. (More information on the dashboard can be found in the
-[Maintaining and Tuning guide][maintaining_tuning].) Typically, this
-catalog duplication rate should be over 90%. If this percentage is
-low, there will be a significantly higher I/O load on the database.
-This load can lead to slowness and resource contention for the
-database and PuppetDB itself.
-
-What Causes Catalogs to Hash Differently?
------
-
-Anything that changes, adds or removes a resource or a relationship
-will cause the catalog for that host to hash differently. Most
-commonly, this is caused by an admin initiated change to a host (i.e.
-adding a new package to the configuration of a web server).
-
-Prior to
-PuppetDB 1.6, this could also be caused by a subtle reordering of
-resources or properties. One example of that scenario is in a two
-puppet master setup, one puppet master would send its resources in a
-different order than the other puppet master. This would cause the
-catalogs to hash differently, even though their content was the same.
-This issue has been fixed in PuppetDB 1.6.
-
-How Can I Detect the Cause of Catalogs Hashing Differently?
------
-
-Unfortunately, diagnosing why a catalog is hashing differently can be
```

Comment:
The user would need more than a SQL query, there's really three pieces to this. The incoming catalog, the stored catalog and the difference between the two. Getting the hash from the stored catalog is easy. Computing the hash for the new catalog is not (i.e. the user would need to invoke one of our functions). The comparison is probably the most difficult aspect of this as we have to hydrate the catalog into the right structures and format the new catalog in the same way. We did this by munging/sorting both sides and spitting to a file. From there the user needed to use diff. Even then it would probably be some help, but could still be difficult to trace down.

I don't think this kind of debugging info is needed anymore. I have not heard of a user having an issue related to de-duplication rate since before 1.6.0. Back then we'd always flush the whole catalog/factset and reinsert everything fresh. Now we're much more targeted, which I think is why the user probably wouldn't even notice a lower duplication rate. I think the users that would, the change would be pretty obvious. You'd have to have a large number of resources changing each time. Compare that with what we had before, which any change (or even accidental change in order of resources) would trigger big I/O on the DB. So the differences now would be much more obvious, before were possibly not even related to a user action.